
// Part One //


General Concepts
1. TCGA is a database of cancer patient data. It contains things like patient information, cancer type, 
mutations, RNA expression, methylation, and more. It is important because its a comprehensive collection 
of genomic data, data that is vital for researchers to understand more about different cancers and their 
treatments. 

2. Some strengths of TCGA include its sheer amount of data, as it contains thousands of patients. It has a wide variety of cancers, allowing researchers to study various types. Some of its weaknesses are that it is a public database, meaning it has less funding than private ones. Also, since it is public, there is a lot of data cleaning that needs to be done before any research can be performed.

Coding Skills
1. To save a file to Github from my laptop, I use the following commands:
    - git add         // Prepares file to be added
    - git commit -m   // Allows me to write message
    - git push        // Puts the file into github
    
2. To use a package, you need to download it and then run it.
    - if (!require("package")) install.packages("package")    // Installs package
    - library(package)                                        // Add package to code
    
3. To use a biodconductor package, you need to these steps:
    - if (!require("BiocManager")) install.packages("BiocManager")    // Installs BiocManager
    - if (!require("package")) BiocManager::install("package")        // Installs bioconductor package
    - library(package)                                                // Adds package to code

4. Boolean indexing labels data values as true or false. This is useful because we can then filter out data easily using ifelse mask statements.

5. survival_mask <- data$vital_status == 'Alive'  // This puts TRUE for alive patients and FALSE for dead
   data[survival_mask, ]                          // Filters out rows of patients that are FALSE (dead)
   
   data$age_category <- ifelse(data$age < 30, 'Young',              // Creates new category called age_category
                      ifelse(data$age > 65, 'Old', 'Middle-Aged'))  // that assigns Young, Old, and Middle-Aged
                                                                    // based on the number in data$age
                                                                    

// Part Two //


```{r setup}
    knitr::opts_knit$set(root.dir = normalizePath("/home1/chengpre/490_cluster/analysis_data")) 
```


```{r Package Setup}
# install, only if needed, and load packages 
if (!require("BiocManager")) install.packages("TCGAbiolinks")
if (!require("BiocManager")) install.packages("BiocManager")
if (!require("SummarizedExperiment")) BiocManager::install("SummarizedExperiment")
if (!require("DESeq2")) BiocManager::install("DESeq2")
if (!require("EnhancedVolcano")) BiocManager::install("EnhancedVolcano")
if (!require("maftools")) install.packages("maftools")
if (!require("ggplot2")) install.packages("ggplot2")
if (!require("survival")) install.packages("survival")
if(!require("survminer")) install.packages("survminer")
if (!require("sesameData")) BiocManager::install("sesameData")
if (!require("sesame")) BiocManager::install("sesame")
if (!require("limma")) BiocManager::install("limma")

library(TCGAbiolinks)
library(SummarizedExperiment)
library(DESeq2)
library(EnhancedVolcano)
library(maftools)
library(ggplot2)
library(survival)
library(survminer)
library(sesameData)
library(sesame)
library(limma)

```


```{r}
# read in clinical data
clin_query <- GDCquery(project = "TCGA-SKCM",
data.category = "Clinical",
data.type = "Clinical Supplement",
data.format = 'BCR Biotab')
# GDCdownload(clin_query)  # Comment out if already downloaded
clinical.BCRtab.all <- GDCprepare(clin_query)

# create various clinical data frames
rad <- clinical.BCRtab.all$clinical_radiation_skcm[-c(1,2),]
drug <- clinical.BCRtab.all$clinical_drug_skcm[-c(1,2),]
clinical <- clinical.BCRtab.all$clinical_patient_skcm
```


```{r}
# read in rna data
rna_query <- GDCquery(project = "TCGA-SKCM",
data.category = "Transcriptome Profiling",
data.type = "Gene Expression Quantification",
workflow.type = "STAR - Counts")
GDCdownload(rna_query)  # Comment out if already downloaded
rna_se <- GDCprepare(rna_query)
```


```{r}
# creates rna_clinical data frame and rna_genes data frame
rna_clinical <- rna_se@colData
rna_clinical <- as.data.frame(rna_clinical)
rna_genes <- rna_se@rowRanges@elementMetadata
rna_genes <- as.data.frame(rna_genes)
rna_counts <- (rna_se@assays@data$unstranded)
rna_counts <- as.data.frame(rna_counts)
```


```{r}
# first need to remove lists
rna_clinical_mask <- colnames(rna_clinical) != "treatments"
rna_clinical <- rna_clinical[ ,rna_clinical_mask]
rna_clinical_mask <- colnames(rna_clinical) != "primary_site"
rna_clinical <- rna_clinical[ ,rna_clinical_mask]
rna_clinical_mask <- colnames(rna_clinical) != "disease_type"
rna_clinical <- rna_clinical[ ,rna_clinical_mask]
```


```{r}
# now save as csv
write.csv(rna_clinical, "/home1/chengpre/490_cluster/analysis_data/skcm_rna_clinical_data.csv", row.names = TRUE)
write.csv(clinical, "/home1/chengpre/490_cluster/analysis_data/skcm_clinical_data.csv", row.names = TRUE)
write.csv(rad, "/home1/chengpre/490_cluster/analysis_data/skcm_rad_data.csv", row.names = TRUE)
write.csv(drug, "/home1/chengpre/490_cluster/analysis_data/skcm_drug_data.csv", row.names = TRUE)
write.csv(rna_genes, "/home1/chengpre/490_cluster/analysis_data/skcm_rna_genes_data.csv", row.names = TRUE)
write.csv(rna_counts, "/home1/chengpre/490_cluster/analysis_data/skcm_rna_counts_data.csv", row.names = TRUE)
```


```{r}
# read in the data
clinical <- read.csv("skcm_clinical_data.csv")
rna_clinical <- read.csv("skcm_rna_clinical_data.csv")
rad <- read.csv("skcm_rad_data.csv")
drug <- read.csv("skcm_drug_data.csv")
rna_genes <- read.csv("skcm_rna_genes_data.csv")
rna_counts <- read.csv("skcm_rna_counts_data.csv")
```




```{r}
# Clean up with only the data I want, "Metastatic" and "Primary solid Tumor"
tumor_mask <- ifelse(rna_clinical$definition == "Metastatic", T, ifelse(rna_clinical$definition == "Primary solid Tumor", T, F))
rna_clinical <- rna_clinical[tumor_mask, ]
rna_counts <- rna_counts[ ,tumor_mask]


# Clean up clinical so that it doesn't have those weird rows at the top
data_mask <- ifelse(clinical$retrospective_collection == "YES", T, ifelse(clinical$retrospective_collection == "NO", T, F))
clean_clinical <- clinical[data_mask, ]
```


```{r}
# creating KM plot

# first clean up "Not Reported" data in vital_status
survival_mask <- rna_clinical$vital_status != "Not Reported"
rna_clinical_clean <- rna_clinical[survival_mask, ]

survival_mask <- clean_clinical$vital_status != "[Not Available]"
clean_clinical <- clean_clinical[survival_mask, ]

# now create survival_time category
clean_clinical$survival_time <- ifelse(clean_clinical$vital_status == "[Not Applicable]", clean_clinical$last_contact_days_to, clean_clinical$death_days_to)

clean_clinical$survival_time <- as.numeric(clean_clinical$survival_time)

# now create death event category
clean_clinical$death_event <- ifelse(clean_clinical$vital_status == "Dead", T, F)

# ensure both dataframes match in length
rna_clinical_clean <- rna_clinical_clean[!duplicated(rna_clinical_clean$patient), ]
missing_barcodes2 <- setdiff(clean_clinical$bcr_patient_barcode, rna_clinical_clean$patient)
missing_barcodes1 <- setdiff(rna_clinical_clean$patient, clean_clinical$bcr_patient_barcode)
rna_clinical_clean <- rna_clinical_clean[!rna_clinical_clean$patient %in% missing_barcodes1, ]
clean_clinical <- clean_clinical[!clean_clinical$bcr_patient_barcode %in% missing_barcodes2, ]


# now create survminer object
survival_object <- Surv(time=clean_clinical$survival_time, event=clean_clinical$death_event)
fit_object <- survfit(survival_object~ rna_clinical_clean$definition, data=rna_clinical_clean)


# create plot
survplot <- ggsurvplot(fit_object , pval=TRUE, ggtheme = 
                        theme(plot.margin = unit(c(1,1,1,1), "cm")), legend = "right")
KM_plot <- survplot$plot + theme_bw() + theme(axis.title =
            element_text(size=20), axis.text = element_text(size=16),
            legend.title = element_text(size=14), legend.text =
            element_text(size=12))

KM_plot
```


```{r}
# Differential Expression

# variable of interest
rna_clinical$definition <- as.factor(rna_clinical$definition)

# remove NA values
na_mask <- !is.na(rna_clinical$race)
rna_clinical <- rna_clinical[na_mask, ]
rna_counts <- rna_counts[ ,na_mask]

na_mask <- ifelse(rna_clinical$race == 'not reported', F, T)
rna_clinical <- rna_clinical[na_mask, ]
rna_counts <- rna_counts[ ,na_mask]

na_mask <- ifelse(rna_clinical$vital_status == 'Not Reported', F, T)
rna_clinical <- rna_clinical[na_mask, ]
rna_counts <- rna_counts[ ,na_mask]


# add covariates into rna_clinical
merged_data <- merge(rna_clinical, drug, by.x = "patient", by.y = "bcr_patient_barcode", all.x = TRUE)
rna_clinical$Radiation <- rna_clinical$patient %in% rad$bcr_patient_barcode
rna_clinical$Chemo <- rna_clinical$patient %in% drug$bcr_patient_barcode & 
                             drug$pharmaceutical_therapy_type == "Chemotherapy"
rna_clinical$Immuno <- rna_clinical$patient %in% drug$bcr_patient_barcode & 
                             drug$pharmaceutical_therapy_type == "Immunotherapy"
rna_clinical$Molecular <- rna_clinical$patient %in% drug$bcr_patient_barcode & 
                             drug$pharmaceutical_therapy_type == "Targeted Molecular therapy"
rna_clinical$Vaccine <- rna_clinical$patient %in% drug$bcr_patient_barcode & 
                             drug$pharmaceutical_therapy_type == "Vaccine"

# covariates
rna_clinical$race <- as.factor(rna_clinical$race)
rna_clinical$gender <- as.factor(rna_clinical$gender)
rna_clinical$vital_status <- as.factor(rna_clinical$vital_status)
rna_clinical$Radiation <- as.factor(rna_clinical$Radiation)
rna_clinical$Chemo <- as.factor(rna_clinical$Chemo)
rna_clinical$Immuno <- as.factor(rna_clinical$Immuno)
rna_clinical$Molecular <- as.factor(rna_clinical$Molecular)
rna_clinical$Vaccine <- as.factor(rna_clinical$Vaccine)

# remove genes < 20
less_20_mask <-  rowSums(rna_counts) >= 20
rna_counts <- rna_counts[less_20_mask, ]
rna_genes <- rna_genes[less_20_mask, ]

# create row and column names
colnames(rna_counts) <- rna_clinical$barcode
row.names(rna_counts) <- rna_genes$gene_id
row.names(rna_clinical) <- rna_clinical$barcode

# match dimensions
common_samples <- intersect(colnames(rna_counts), rownames(rna_clinical))
rna_counts <- rna_counts[ ,common_samples]
rna_clinical <- rna_clinical[common_samples, ]

# dimensions
dim(rna_counts)
dim(rna_genes)
dim(rna_clinical)

rna_counts <- as.matrix(rna_counts)
```

```{r}
# one-hot encoding

rna_clinical$Radiation <- as.logical(rna_clinical$Radiation)
rna_clinical$Radiation <- as.integer(rna_clinical$Radiation)

rna_clinical$Chemo <- as.logical(rna_clinical$Chemo)
rna_clinical$Chemo <- as.integer(rna_clinical$Chemo)

rna_clinical$Immuno <- as.logical(rna_clinical$Immuno)
rna_clinical$Immuno <- as.integer(rna_clinical$Immuno)

rna_clinical$Molecular <- as.logical(rna_clinical$Molecular)
rna_clinical$Molecular <- as.integer(rna_clinical$Molecular)

rna_clinical$Vaccine <- as.logical(rna_clinical$Vaccine)
rna_clinical$Vaccine <- as.integer(rna_clinical$Vaccine)


```




```{r}
# create DESeq2 
dds <- DESeqDataSetFromMatrix(countData = rna_counts,
                              colData = rna_clinical,
                              design=~ race + gender + vital_status + Radiation + Chemo + Immuno + Molecular +                                        Vaccine + definition)

dds_obj <- DESeq(dds)

resultsNames(dds_obj)

results <- results(dds_obj, format = "DataFrame", contrast = c("definition", "Metastatic" , "Primary solid Tumor"))

results <- data.frame(results)

results$gene_name <- rna_genes$gene_name
results$"-log10(padj)" <- -log10(results$padj)
```


```{r}
# create volcano plot

write.csv(DESeq, "/home1/chengpre/490_cluster/analysis_data/skcm_DESeq_data.csv", row.names = TRUE)

EnhancedVolcano(results,
                lab = results$gene_name,
                x = 'log2FoldChange',
                y = '-log10(padj)',
                title = 'Sample Definition: Tumor vs Normal Tissue',
                pointSize = 1.0,
                labSize = 5.0)
```





```{r}
# read in methylation data
query <- GDCquery(project = "TCGA-SKCM",
                  data.category = "DNA Methylation",
                  data.type = "Methylation Beta Value",
                  platform = "Illumina Human Methylation 450",
                  data.format = "Dataframe",
                  access = "open")
#GDCdownload(query, files.per.chunk=2)
methylation450 <- GDCprepare(query)
```


```{r}
# create data frames
methylation_clinical <- as.data.frame(methylation450@colData)
betas <- as.data.frame(methylation450@assays@data@listData)
cpg_sites <- as.data.frame(methylation450@rowRanges@elementMetadata)
```


```{r}
# save as csv
column_mask <- ifelse(colnames(methylation_clinical) %in% c('treatments', 'primary_site', 'disease_type'), F, T)
methylation_clinical <- methylation_clinical[,column_mask]
write.csv(methylation_clinical, 'skcm_methylation_clinical.csv')


site_mask <- !grepl('-', cpg_sites$gene) & !grepl(';', cpg_sites$gene) & !is.na(cpg_sites$gene) & complete.cases(betas)
betas <- betas[site_mask,]
cpg_sites <- cpg_sites[site_mask,]

write.csv(betas, 'skcm_methylation_betas.csv')
write.csv(cpg_sites, 'skcm_cpg_sites.csv')
```


```{r}
# read in csvs
methylation_clinical <- read.csv("skcm_methylation_clinical.csv")
betas <- read.csv("skcm_methylation_betas.csv", row.name=1)
cpg_sites <- read.csv("skcm_cpg_sites.csv")
```


```{r}
# new tumor column
normal_tissue_mask <- methylation_clinical$definition == "Solid Tissue Normal"
methylation_clinical$tumor <- !normal_tissue_mask


# create linear fit
mval <- t(apply(betas, 1, function(x) log2(x/(1-x))))

design <- model.matrix(~ tumor, data = methylation_clinical)
fit <- lmFit(mval, design)
fit2 <- eBayes(fit)
```


```{r}
# create data frame
dat <- data.frame(foldchange = fit[["coefficients"]][,2], logPvalue = -log10(p.adjust(fit2[["p.value"]][,2],
                  method='BY')), geneName = cpg_sites$gene)
dat$threshold <- as.factor(abs(dat$foldchange) < 1)


# visualization
cols <- c("TRUE" = "grey", "FALSE" = "blue")
ggplot(data=dat, aes(x=foldchange, y = logPvalue, color=threshold)) +
  geom_point(alpha=.2, size=0.6) +
  scale_colour_manual(values = cols) +
  geom_vline(xintercept = 1, colour="#990000", linetype="dashed") + 
  geom_vline(xintercept = - 1, colour="#990000", linetype="dashed") +
  geom_hline(yintercept = 2, colour = "#990000", linetype="dashed") +
  theme(legend.position="none") +
  xlab("Fold Change") +
  ylab("-log10 p value") +
  theme_bw() +
  theme(legend.position = "none")
```

```{r}






```









   